eggplant/market/models/cart.py

```diff
@@ -53,35 +60,49 @@ def add_to_items(self, product=None, quantity=1, delivery_date=None):
 
     def remove_from_items(self, product=None, quantity=1, delivery_date=None):
         current = self.items.filter(product=product,
-                                    delivery_date=delivery_date)
-        if current.exists():
-            if current[0].quantity - quantity > 1:
-                current.update(quantity=models.F('quantity') - quantity)
+                                    delivery_date=delivery_date).first()
+        if current:
+            if current.quantity - quantity > 1:
+                current.quantity -= quantity
+                current.save()
             else:
                 self.items.filter(basket=self, product=product, delivery_date=delivery_date).delete()
 
     def get_total_amount(self):
-        # TODO: This should return a Money type instead of decimal
-        # TODO: When calculating this, we should fail when currencies differ
-        total = Decimal('0')
-        for item in self.items.all():
-            total += item.quantity * item.product.price.amount
-        return total
+
+        # FIXME this check should be removed, it is better to check for the currenices when an item is added
+        # sanity check for the currencies
+
+        if 1 != self.items.all().values('product__price_currency').annotate(total=Count('product__price_currency')).count():
+            raise ValidationError('The products have different currencies')
+
+        total = self.items.all().aggregate(total=Sum(F('product__price') * F('quantity')), cur=Max('product__price_currency'))
+        return Money(total['total'], total['cur'])
 
     def get_items_count(self):
         return self.items.all().count()
 
     @transaction.atomic
     def do_checkout(self):
-        # TODO: Why on earth are we importing this here
-        from .payment import Payment
-        Payment.objects.create(
+        payment = Payment.objects.create(
             amount=self.get_total_amount(),
+            # FIXME: get valid account
```

Comment:
Well in ideal case: a user should be created with a personal account (better name would be organization). It should be a trigger and account model should have one more field - personal. 
So every user has own default account. Later on the user can join other non personal accounts.

And somewhere in navigation bar he can select which account is he using (probably a session variable). Basket should be based on account_id. This would allow to share the basket for the whole organization. Anyway it has nothing to do with what I have done.